risc0/cargo-risczero/src/commands/build_guest.rs

```diff
@@ -0,0 +1,201 @@
+            .root_package()
+            .context("failed to parse Cargo.toml")?
+            .name;
+        eprintln!("Building the riscv32im-risc0-zkvm-elf binary for {pkg_name}...");
+        ensure_binary("docker", &["--version"])?;
+        if let Err(err) = self.check_cargo_lock() {
+            eprintln!("{err}");
+        }
+        let package_name = pkg_name.replace('-', "_");
+        self.create_dockerfile(package_name.as_str())?;
+        self.build()?;
+        self.clean()?;
+
+        let paths = fs::read_dir(&format!("./elfs/{package_name}/"))?;
+        println!("ELFs ready at:");
+        for path in paths {
+            let entry = path.unwrap().path();
+            println!("{} - ImageID: {}", entry.display(), self.image_id(&entry)?);
+        }
+
+        Ok(())
+    }
+
+    /// Create the dockerfile.
+    ///
+    /// Overwrites if a dockerfile already exists.
+    fn create_dockerfile(&self, pkg_name: &str) -> Result<()> {
+        let manifest_env = &[("CARGO_MANIFEST_PATH", self.manifest_path.as_str())];
+        let pkg_env = &[("PKG_NAME", pkg_name)];
+        let c_flags = format!(
+            "-C passes=loweratomic -C link-arg=-Ttext=0x{:08X} -C link-arg=--fatal-warnings",
+            memory::TEXT_START
+        );
+        let c_flags_env = &[("RUSTFLAGS", c_flags.as_str())];
+
+        let build = DockerFile::new()
+            .from_alias("build", "risczero/risc0-guest-builder:v0.17")
+            .workdir("/src")
+            .copy(".", ".")
+            .env(manifest_env)
+            .env(c_flags_env)
+            .run(
+                "cargo +risc0 fetch --target riscv32im-risc0-zkvm-elf --manifest-path $CARGO_MANIFEST_PATH --locked",
+            )
+            .run(
+                "CARGO_TARGET_DIR=target \\\n\
+                \tcargo +risc0 build \\\n\
+                \t--locked \\\n\
+                \t--release \\\n\
+                \t--target riscv32im-risc0-zkvm-elf \\\n\
+                \t--manifest-path $CARGO_MANIFEST_PATH")
+            .run(r"find target/riscv32im-risc0-zkvm-elf/release -maxdepth 1 -type f -exec test -x {} \; -exec cp {} /tmp/ \;");
+
+        let binary: DockerFile<'_> = DockerFile::new()
+            .comment("binary stage")
+            .from_alias("binary", "scratch")
+            .env(pkg_env)
+            .copy_from("build", "/tmp", "/$PKG_NAME");
+
+        let file = DockerFile::new().dockerfile(build).dockerfile(binary);
+        let mut dockerfile = File::create("Dockerfile")?;
+        dockerfile.write_all(file.to_string().as_bytes())?;
+        let mut dockerignore = File::create(".dockerignore")?;
+        dockerignore.write_all(DOCKER_IGNORE.as_bytes())?;
+
+        Ok(())
+    }
+
+    /// Build the dockerfile and ouputs the ELF.
+    ///
+    /// Overwrites if an ELF with the same name already exists.
+    fn build(&self) -> Result<()> {
+        Command::new("docker")
+            .args(["build", "--output=elfs/", "--target=binary", "."])
```

Comment:
I think the output should land under `target/`, so maybe `target/riscv-guest/riscv32im-risc0-zkvm-elf/docker`?

Reply:
modified 👍